File: apps/mobile/lib/translation/translation_presentation.dart
import 'translation_domain.dart';

class TranscriptSegmentIdentity {
  const TranscriptSegmentIdentity(this.streamId, this.segmentId);

  final String? streamId;
  final String segmentId;

  @override
  bool operator ==(Object other) =>
      other is TranscriptSegmentIdentity &&
      other.streamId == streamId &&
      other.segmentId == segmentId;

  @override
  int get hashCode => Object.hash(streamId, segmentId);
}

class LiveTranscriptSegment {
  const LiveTranscriptSegment({
    required this.streamId,
    required this.segmentId,
    required this.text,
    required this.isFinal,
  });

  final String? streamId;
  final String segmentId;
  final String text;
  final bool isFinal;

  TranscriptSegmentIdentity get identity =>
      TranscriptSegmentIdentity(streamId, segmentId);
}

class TranslationPresentation {
  const TranslationPresentation({
    required this.utterances,
    required this.liveSpeechSegments,
  });

  final List<TranslationUtterance> utterances;
  final List<LiveTranscriptSegment> liveSpeechSegments;
}

TranslationPresentation buildTranslationPresentation(
  List<LiveTranscriptSegment> segments,
  List<TranslationUtterance> utterances,
) {
  final claimed = <TranscriptSegmentIdentity>{};
  for (final utterance in utterances) {
    for (final segmentId in utterance.sourceSegmentIds) {
      claimed.add(
        TranscriptSegmentIdentity(utterance.identity.streamId, segmentId),
      );
    }
  }
  return TranslationPresentation(
    utterances: List.unmodifiable(utterances),
    liveSpeechSegments: List.unmodifiable(
      segments.where(
        (segment) => !segment.isFinal || !claimed.contains(segment.identity),
      ),
    ),
  );
}

class BilingualTranscriptBlockView {
  const BilingualTranscriptBlockView({
    required this.sourceLabel,
    required this.sourceText,
    required this.translationLabel,
    required this.translationText,
    required this.status,
    required this.secondaryHint,
  });

  final String sourceLabel;
  final String sourceText;
  final String translationLabel;
  final String translationText;
  final TranslationStatus status;
  final String? secondaryHint;
}

BilingualTranscriptBlockView buildBilingualTranscriptBlockView(
  TranslationUtterance utterance,
) {
  final translationText = switch (utterance.status) {
    TranslationStatus.pending => 'Translating...',
    TranslationStatus.finalResult => utterance.translatedText ?? '',
    TranslationStatus.failed => 'Translation unavailable',
  };
  return BilingualTranscriptBlockView(
    sourceLabel: 'ORIGINAL · ${utterance.sourceLanguage.toUpperCase()}',
    sourceText: utterance.sourceText,
    translationLabel:
        'TRANSLATION · ${utterance.targetLanguage.code.toUpperCase()}',
    translationText: translationText,
    status: utterance.status,
    secondaryHint: utterance.status == TranslationStatus.failed
        ? 'Original transcript is still available.'
        : null,
  );
}

const translationUnavailableWarning =
    'Translation is unavailable. Original transcript will continue.';

File: apps/web/lib/realtime/translation-presentation.ts
import type { TranscriptSegment } from "./transcript.ts";
import type {
  TargetLanguage,
  TranslationSessionErrorState,
  TranslationState,
  TranslationUtteranceState,
} from "./translation.ts";

export type TranslationPresentation = {
  utterances: readonly TranslationUtteranceState[];
  liveSegments: readonly TranscriptSegment[];
};

export type BilingualTranscriptBlockView = {
  sourceLabel: string;
  sourceText: string;
  translationLabel: string;
  translationText: string;
  status: TranslationUtteranceState["status"];
  secondaryHint: string | null;
};

const LANGUAGE_LABELS: Record<"vi" | TargetLanguage, string> = {
  vi: "Vietnamese",
  en: "English",
  ja: "Japanese",
  ko: "Korean",
  "zh-CN": "Chinese (Simplified)",
  th: "Thai",
  fr: "French",
  de: "German",
  es: "Spanish",
};

export function transcriptSegmentKey(
  streamId: string | undefined,
  segmentId: string,
): string {
  return `${streamId ?? ""}\u0000${segmentId}`;
}

export function buildTranslationPresentation(
  segments: readonly TranscriptSegment[],
  utterances: readonly TranslationUtteranceState[],
): TranslationPresentation {
  const claimedSegments = new Set(
    utterances.flatMap((utterance) =>
      utterance.sourceSegmentIds.map((segmentId) =>
        transcriptSegmentKey(utterance.streamId, segmentId),
      ),
    ),
  );
  return {
    utterances,
    liveSegments: segments.filter(
      (segment) =>
        segment.kind === "interim" ||
        !claimedSegments.has(
          transcriptSegmentKey(segment.streamId, segment.id),
        ),
    ),
  };
}

export function getTranslationDisplayText(
  utterance: TranslationUtteranceState,
): string {
  if (utterance.status === "pending") {
    return "Translating...";
  }
  if (utterance.status === "failed") {
    return "Translation unavailable";
  }
  return utterance.translatedText ?? "";
}

export function buildBilingualTranscriptBlockView(
  utterance: TranslationUtteranceState,
): BilingualTranscriptBlockView {
  return {
    sourceLabel: `ORIGINAL · ${utterance.sourceLanguage.toUpperCase()}`,
    sourceText: utterance.sourceText,
    translationLabel: `TRANSLATION · ${utterance.targetLanguage.toUpperCase()}`,
    translationText: getTranslationDisplayText(utterance),
    status: utterance.status,
    secondaryHint:
      utterance.status === "failed"
        ? "Original transcript is still available."
        : null,
  };
}

export function getLanguageLabel(
  language: "vi" | TargetLanguage,
): string {
  return LANGUAGE_LABELS[language];
}

export function getTranslationSessionWarning(
  errors: readonly TranslationSessionErrorState[],
  activeStreamId: string | null,
): string | null {
  const relevantError =
    activeStreamId === null
      ? errors.at(-1)
      : errors.find((error) => error.streamId === activeStreamId);
  return relevantError === undefined
    ? null
    : "Translation is unavailable. Original transcript will continue.";
}

export function shouldUseBilingualPresentation(
  translationExpected: boolean,
  state: TranslationState,
): boolean {
  return (
    translationExpected ||
    state.configurations.length > 0 ||
    state.utterances.length > 0 ||
    state.sessionErrors.length > 0
  );
}
